src/PointwiseFunctions/Hydro/ComovingMagneticField.cpp

```diff
@@ -0,0 +1,71 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/Hydro/ComovingMagneticField.hpp"
+
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+/// \cond
+namespace hydro {
+template <typename DataType, size_t Dim, typename Fr>
+tnsr::A<DataType, Dim, Fr> comoving_magnetic_field(
+    const tnsr::I<DataType, Dim, Fr>& eulerian_b_field,
+    const tnsr::I<DataType, Dim, Fr>& transport_velocity,
+    const tnsr::i<DataType, Dim, Fr>& spatial_velocity_oneform,
+    const Scalar<DataType>& lorentz_factor,
+    const Scalar<DataType>& lapse) noexcept {
+  auto result = make_with_value<tnsr::A<DataType, Dim, Fr>>(lapse, 0.0);
+
+  get<0>(result) =
+      get(lorentz_factor) *
+      get(dot_product(eulerian_b_field, spatial_velocity_oneform)) / get(lapse);
```

Comment:
Could you please eliminate the allocation in the call to `dot_product` by using the `not_null` version? You should be able to do this by creating a `Scalar<DataVector>` where the `DataVector` aliases with the first element in `result`. If the functions you're adding are doing allocations then it'll severely limit where we can actually call them, unfortunately.